### DISCUSSION TITLE
Loading widget trees via string

### DISCUSSION THREAD

**Discussion by tomplast** [beeware/toga#2401] [Ideas]
I don't know if there's any interest - except from me - but I've created a library for loading Toga widget trees via strings. So instead of having to create all the widgets in code and appending them to each other I can just do it in the following manner:

```
from togawil import load_widget_from_string
load_widget_from_string("""
Box:
    id: "box1"
    Button:
        id: "button1"
""")
```

This will return a Box widget (with id box1) and a Button (with id button1) as a child.

It also supports specifying event handlers:

```
def button_handler(widget):
    print('hello world')

from togawil import load_widget_from_string
widget = load_widget_from_string("""
Box:
    id: "box1"
    Button:
        id: "button1"
        on_press: button_handler
""")
```

As I've said I don't know if this is useful except me but I like the idea of declaring widgets in this way and if you do too, enjoy :). If my motivation won't die and there's interest I'm planning on implementing templating and other features.

Oh, I almost forgot.. The library can be found at https://github.com/tomplast/togawil

**Comment by freakboy3742** [comment]
What you're describing is #2293. The feature has been requested many times; as is noted on that ticket, the devil is in the detail. 

You've discovered the easy part - that hierarchical markup languages map well to Toga's widget layout. You've used YAML; but JSON, XML, or any other markup language could be used here. What you haven't explored is the hard part - questions like: 
1. How do you map markup values to constructor arguments (e.g., how do you define a Selection widget's list of choices, or the min/max value of a Slider) 
2. Does your markup language require an explicit definition of a mapping for every widget, or can a mapping be implied based on the "shape" of the markup language?
3. How does this widget creation process with overall app lifecycle (e.g., how do you define an app with multiple windows?)

As has been noted on #2293 - I definitely encourage you to dig into this further if you're motivated by markup languages as a use case. As you've already discovered, this may be able to develop as a third party library; but if you've got ideas about how this could integrate with the core app lifecycle, we're happy to entertain that discussions.

The one request I'd make is that you avoid the "toga" prefix for any package you publish to PyPI - we'd like to reserve that name for official packages, and encourage community contributions to use the `togax-` prefix instead.

**Reply by tomplast** [reply]
Hi @freakboy3742 

1. I send the markup values to the constructor of the widget as named arguments and (since a few minutes ago) I cast the values to int / float (supported at the moment) based on the type hinting for that particular widget's constructor. Example for a Slider:

```
def my_callback(slider):
    print(slider.value)

def build(app):
    widget_string = '''
Box:
    Slider:
        min: -5.2
        max: 10
        value: 7
        on_change: my_callback
    '''
    widget = load_widget_from_string(widget_string)
    return widget
```

3. I'm not certain I understand your question but for every widget you want you'll have to specify the name of it and the required arguments (for that particular class).

4. I have not thought about / included any of that in the scope. Right now it just returns a widget instance (with optional children).

...And as per your request I have renamed my library to togax-wil.
